dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYamlHelper.java

```diff
@@ -187,6 +213,152 @@ private static DockstoreYaml12 readDockstoreYaml12(final String content) throws
+                }
+                throw new DockstoreYamlException(errorMessage);
+            }
+        }
+    }
+
+    /**
+     * Finds a valid .dockstore.yml property that is the closest to the unknown property according to its Levenshtein distance.
+     * @param dockstoreYamlClass The DockstoreYaml class. Allowed values: DockstoreYaml12.class, Dockstore11.class
+     * @param unknownProperty The unknown property to find a suggestion for.
+     * @return A suggested property.
+     */
+    public static String getSuggestedDockstoreYamlProperty(Class<? extends DockstoreYaml> dockstoreYamlClass, String unknownProperty) {
+        Set<String> validProperties = getDockstoreYamlProperties(dockstoreYamlClass);
+        LevenshteinDistance levenshteinDistance = new LevenshteinDistance();
+        int shortestDistance = Integer.MAX_VALUE;
+        String shortestDistanceProperty = "";
+
+        if (validProperties.contains(unknownProperty) || unknownProperty.isEmpty()) {
+            return unknownProperty;
+        }
+
+        for (String validProperty : validProperties) {
+            int distance = levenshteinDistance.apply(unknownProperty, validProperty);
+            if (distance < shortestDistance) {
+                shortestDistance = distance;
+                shortestDistanceProperty = validProperty;
+            }
+        }
+
+        // Return the property if the number of changes needed to be made is less than the length of the unknown property.
+        // This is to prevent suggestions that don't make sense.
+        if (shortestDistance < unknownProperty.length()) {
+            return shortestDistanceProperty;
+        } else {
+            return "";
+        }
+    }
+
+    /**
+     * Gets all the properties of a .dockstore.yml as a list of strings.
+     * The convention that makes this work is that all the property fields in the DockstoreYaml classes are private.
+     * Uses bread-first search to find all the classes.
+     * @param dockstoreYamlClass The DockstoreYaml class. Allowed values: DockstoreYaml12.class, DockstoreYaml11.class
+     * @return A set of properties belonging to the .dockstore.yml version
+     */
+    public static Set<String> getDockstoreYamlProperties(Class<? extends DockstoreYaml> dockstoreYamlClass) {
+        Set<String> properties = new HashSet<>();
+        Queue<Class> dockstoreYmlPropertiesQueue = new ArrayDeque<>(); // A queue to process the property classes
+        List<Class> discoveredClasses = new ArrayList<>();
+        final String dockstoreYamlPackageName = dockstoreYamlClass.getPackageName();
+
+        discoveredClasses.add(dockstoreYamlClass);
+        dockstoreYmlPropertiesQueue.add(dockstoreYamlClass); // Add the high level class to the queue
+
+        while (!dockstoreYmlPropertiesQueue.isEmpty()) {
+            Class propertyClass = dockstoreYmlPropertiesQueue.poll(); // Get a class in the queue to process
```

Comment:
Micro-optimization: reflection is relatively expensive in the JVM, so you could have static fields with the results of these. In practice, I don't think it will matter, so probably not worth changing.